File: dart/word-count/lib/word_count.dart
class WordCount {
  final regex = RegExp(r"(\w+'?\w+|\d+|\w+)");

  Map countWords(String phrase) =>
      splitToList(phrase).fold<Map<String, int>>({}, updateCount);

  Map<String, int> updateCount(Map<String, int> result, String word) =>
      result..update(word, (val) => ++val, ifAbsent: () => 1);

  Iterable<String> splitToList(String phrase) =>
      regex.allMatches(phrase.toLowerCase()).map((match) => match.group(0));
}

File: typescript/word-count/word-count.ts
export default class Words {
  private static regex = new RegExp(/\s+|\t|\n/g);
  count(phrase: string): object {
    const result = new Map();
    this.processPhrase(phrase).forEach((word) =>
      result.set(word, (result.get(word) || 0) + 1)
    );
    return result;
  }

  processPhrase = (phrase: string) =>
    phrase
      .toLowerCase()
      .split(Words.regex)
      .filter((word) => word.length > 0);
}
